### PR TITLE
Feature/auth

### PR DESCRIPTION
**Pull request by Matt-Guest**
### What

* Added authentication for the internal services
* GET /instances Will return a list
  * includes query param to filter on state
* Lock the job state when set to 'submitted' only an authenticated request can update the job
* PUT /instances/{instance_id}/inserted_observations/{inserted_observations} update inserted observations
* Fixed a bug where by submitting a job that it didn't update the instance

### How to review

Review changes, Test and read the swagger spec

### Who can review

Anyone

### CODE REVIEW

File: models/models.go [hunk 1]
```diff
@@ -13,7 +13,7 @@ type Job struct {
 	Recipe        string          `json:"recipe,omitempty"`
 	State         string          `json:"state,omitempty"`
 	UploadedFiles *[]UploadedFile `json:"files,omitempty"`
-	Links         JobLinks        `json:"links,omitempty"`
```
Comment (eldeal) [inline models/models.go]: Does this mean the JobLinks struct is no longer needed?

File: postgres/postgres.go [hunk 2]
```diff
@@ -196,10 +224,37 @@ func (ds Datastore) GetInstance(host, instanceID string) (models.Instance, error
...
 
+// GetInstances from postgres
+func (ds Datastore) GetInstances(host string, filter []string) ([]models.Instance, error) {
+	if len(filter) == 0 {
+		filter = allFilterStates
+	}
+	var instances []models.Instance
+	rows, err := ds.getInstances.Query(pg.Array(filter))
+	if err != nil {
+		return []models.Instance{}, err
```
Comment (eldeal) [inline postgres/postgres.go]: Return nil instead of empty [] might be cleaner

File: postgres/postgres_test.go [hunk 3]
```diff
@@ -54,6 +54,22 @@ func TestGetInstance(t *testing.T) {
 	})
 }
 
+func TestGetInstances(t *testing.T) {
+	t.Parallel()
+	Convey("When a request for all, a list of instances are returned", t, func() {
+		jsonContent := "{ \"state\":\"Created\"}"
```
Comment (eldeal) [inline postgres/postgres_test.go]: Might not be something we need to fix right now, but job states should all be lower case. As long as it's consistent it's not a problem for now, but definitely one to put on the list to sort soon.

File: api/api.go [hunk 4]
```diff
@@ -183,10 +189,40 @@ func (api *ImportAPI) updateJob(w http.ResponseWriter, r *http.Request) {
...
-		log.Info("import job was queued", log.Data{"job_id": jobID})
+		log.Info("import job was queued", log.Data{"job": job, "job_id": jobID, "is_auth": isAuth})
 	}
 }
 
+func (api *ImportAPI) getInstances(w http.ResponseWriter, r *http.Request) {
+	filtersQuery := r.URL.Query().Get("instance_states")
+	var filterList []string
+	if filtersQuery == "" {
+		filterList = nil
```
Comment (Matt-Guest) [inline api/api.go]: Good spot!

File: api/api.go [hunk 5]
```diff
@@ -323,28 +359,47 @@ func (api *ImportAPI) getDimensionValues(w http.ResponseWriter, r *http.Request)
...
-	log.Info("added node id", log.Data{"instance_id": instanceID, "name": dimensionName, "nodeValue": nodeValue})
+	log.Info("added node id", log.Data{"instance_id": instanceID, "dimension": dimension})
+}
+
+func (api *ImportAPI) addInsertedObservations(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	instanceID := vars["instance_id"]
+	insertedObservations, err := strconv.Atoi(vars["inserted_observations"])
+	if insertedObservations == 0 || err != nil {
+		log.Error(errors.New("Missing / invalid parameters"), log.Data{"instance_id": instanceID, "inserted_observations": insertedObservations})
```
Comment (Matt-Guest) [inline api/api.go]: Done

File: api/authentication.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
+package api
+
+import (
+	"net/http"
+	"github.com/ONSdigital/go-ns/log"
+	"errors"
+)
+
+// Authenticator A simple authentication method. This will be replaced in the future, after a thin-slices has been delivered
```
Comment (Matt-Guest) [inline api/authentication.go]: done

File: api/authentication.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+	return Authenticator{ secretKey: key, headerName: headerName}
+}
+
+// Check wraps a HTTP handle. If authentication fails an error code is returned else the HTTP handler is called
+func (a *Authenticator) Check(handle func(http.ResponseWriter, *http.Request)) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		key := r.Header.Get(a.headerName)
+		if key == "" {
+			http.Error(w, "No authentication header provided", http.StatusForbidden)
+			log.Error(errors.New("Client missing token"), log.Data{"header":a.headerName})
```
Comment (Matt-Guest) [inline api/authentication.go]: done

File: models/models.go (hunk 1 above)
Comment (Matt-Guest) [inline models/models.go]: Good spot! I removed it

File: postgres/postgres.go (hunk 2 above)
Comment (Matt-Guest) [inline postgres/postgres.go]: Done, also cleaned up the other functions

File: postgres/postgres_test.go (hunk 3 above)
Comment (Matt-Guest) [inline postgres/postgres_test.go]: I changes the tests to lower case, I think we need to force lower case characters when a request comes in. I Added a TODO

**Pull request merged**